test-control/Test/Control/RefCount.hs

```diff
@@ -80,3 +82,25 @@ prop_refCount = once $ ioProperty $ do
 #else
     check = \case Left (AssertionFailed _) -> False; Right () -> True
 #endif
+
+prop_removeReferenceN :: Positive Int -> NonNegative Int -> Property
+prop_removeReferenceN (Positive n) (NonNegative m) = ioProperty $ do
+    obj <- newMVar False
+    ref <- unsafeMkRefCounterN (RefCount n) $ Just (void $ modifyMVar_ obj (\x -> pure (not x)) )
+
+    e1 <- try $ removeReferenceN ref (fromIntegral m)
+    n1 <- readRefCount ref -- 0
+    b1 <- readMVar obj -- True
+
+    pure $
+        counterexample "e1" (if n < m || m == 0
+                             then check e1
+                             else isRight e1) .&&.
+        counterexample "n1" (n1 == RefCount (n - m)) .&&.
+        counterexample "b1" (b1 == (m >= n))
+  where
+#ifdef NO_IGNORE_ASSERTS
+    check = \case Left (AssertionFailed _) -> True; Right () -> False
+#else
+    check = \case Left (AssertionFailed _) -> False; Right () -> True
+#endif
```

Comment:
This property currently fails, because the behaviour is different with assertions turned on or off. Should be relatively straightforward to fix here, or to change `removeReferenceN` slightly

Reply:
Fixed now